VSR: Remain available despite asymmetric partitions

sentientwaffle opened the pull request:
(This design is already documented in `vsr/README.md`.)

## Summary:

1. Primary sends `command=commit` heartbeats to tell backups that the primary is alive.
2. Primary pauses sending `command=commit` heartbeats if it stops receiving `prepare_ok`s.
3. If the backups think the primary isn't alive, they start a SVC.
4. If a replica collects a VC quorum of SVCs, they start a DVC.
5. Any replica which receives a DVC will jump views and send its own DVC (even if they did not participate in the SVC).

Previously a partitioned replica would endlessly cycle through view-changes.
If the partition is asymmetric, this could disrupt the remaining (otherwise healthy) cluster.
But now in that scenario the isolated replica would not complete an SVC quorum.

## Also included:
- Split `ping`/`ping_client`, `pong`/`pong_client`.
- `prepare_timeout`
    - Previously, this would retry the first prepare in the pipeline if it does not have a quorum of acks, and otherwise do nothing.
    - Now, this retries the first prepare in the pipeline without a quorum of acks.
- Simulate asymmetric partitions in the VOPR's `PacketSimulator`.
    - (This is of limited use since the VOPR doesn't really test liveness.)

Aside: A quirk of these SVC changes: when `replica_count=2`, I think the only way to change views when is now via crashes.
Which was surprising to me but makes sense. Both the replication and view-change quorums are 2.
So to start a view change requires a quorum of 2 SVCs, but the primary doesn't send SVCs (since if the backups could hear its SVCs, then it may as well stay primary anyway).
I think this is fine, just funny/surprising.



## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance. (This will change performance in certain conditions, e.g. network partitions, but that is not measured by the benchmark).

sentientwaffle committed:
VSR: Asymmetric partition-tolerance

Summary: Only initiate a view-change (increment the view number) when a view-change quorum wants one. Voting for "replica wants a view-change" can now take place in `status=normal` and/or `status=view_change`.

Right now, the view-change flow is:

  1. Replica decides it wants to start a new view.
  2. Replica sends SVC to all replicas.
  3. Upon receiving an SVC, a replica sends an SVC to all replicas.
  4. Upon receiving a quorum of SVCs, a replica sends a DVC to the primary of new new view.
  5. Upon receiving a quorum of DVCs, the primary of the new view begins repair.
  6. New primary sends SV to all replicas.

The above sequence is based directly on VRR's view-change protocol (https://pmg.csail.mit.edu/papers/vr-revisited.pdf),
which is designed (in concert with Recovery Protocol) to be used without persistent storage.
But the original Viewstamped Replication protocol was designed for persistent storage (no recovery protocol), which allows the SVC step to be skipped.

Without the SVC step, the flow is:

  1. Replica decides it wants to start a new view.
  2. Replica sends DVC to all replicas.
  3. Upon receiving an DVC, a replica sends an DVC to all replicas.
  4. Upon receiving a quorum of DVCs, the primary of the new view begins repair. (Only the primary collects the DVCs — backups just use them to trigger a view jump).
  5. New primary sends SV to all replicas.

However, neither of these tolerate asymmetric partitions.
For example, if a particular replica can send to the remainder of the cluster, but cannot receive anything, it will continually send SVCs (former case) or DVCs (latter case), which compromise cluster liveness by repeatedly initiating view-changes.

This commit implements the following hybrid strategy:

- Modify `normal_status_timeout`:
  - Run on both primary + backup.
  - When the timeout fires start a `start_view_change_message_timeout` and send an SVC message to all replicas (including ourself) (while still in `status=normal`).
  - (Instead of incrementing the view number before sending an SVC, just include the replica's current view.)
- When `start_view_change_message_timeout` times out, send an SVC message to all replicas (which includes the replica's current `view`).
- When the replica commits:
  - `normal_status_timeout.reset()`
  - `start_view_change_message_timeout.stop()`
  - reset the SVC quorum
- `replica.on_start_view_change`:
  - Collect the SVC for a quorum. (Without changing our status — we might be `status=normal` or `status=view_change`).
  - When the SVC quorum is complete, switch to `status=view_change` and start sending DVC messages for "1 + the highest view of any SVC".
- When a replica receives a DVC, transitions to `view_change` and can send a DVC immediately (i.e. w/o waiting for an SVC quorum).
- The primary can have a timeout which prepares a noop message periodically if nothing else was prepared in the mean time, to keep the SVC timeout from going off when the network is actually healthy.

This means that if a `status=normal` replica is partitioned (either symmetric or asymmetric) from the remainder of the (healthy) cluster, and then the partition heals, the replica resumes in the exact same view that it left, without ever transitioning to view-change status.

Aside: The behavior of `normal_status_timeout` and `view_change_status_timeout` is the same, but they are separate because we might want their timeout (i.e. the number of ticks) to differ.

sentientwaffle committed:
PacketSimulator: Simulate asymmetric partitions

Not super useful right now because the VOPR can't detect transient liveness bugs, but oh well.

sentientwaffle committed:
VSR: Don't send SVC while committing

If a commit takes an usually long time, starting an SVC wouldn't actually help.

sentientwaffle committed:
VSR: Use pings instead of commits to detect partition

Use pings/pongs as heartbeats, instead of commit-heartbeats, to detect partitions (i.e. to trigger SVC).

Originally this was going to be handled by the primary periodically preparing a noop message, and then starting an SVC if we haven't committed anything in a while.

Commit-heartbeats would work, but there are a few downsides (in descending priority):
- In reconfiguration, the old cluster should stop preparing messages.
- The ReplySequencer implementation would become tricker and much less efficient, since it wouldn't be able to use op numbers directly.
- Less efficient.

sentientwaffle committed:
VSR: ping_client/pong_client

Use `ping`/`pong` between replicas.
Use `ping_client`/`pong_client` between client-replica.

Replica-replica and replica-client ping/pongs are similar, but serve different purposes.
The messages have different invariants — in some cases this is quite subtle.
And we have to be extra careful since a message intended for a replica can be misdirected to a client (or vice-versa).
To guard against footguns, split them into separate message types.

sentientwaffle committed:
VSR: s/commit_timeout/commit_message_timeout

Disambiguates "commit" (noun) from "commit" (verb).

sentientwaffle committed:
VSR: Refactor on_ping

sentientwaffle committed:
VSR: CheckQuorum → primary abdication

If the primary goes too long without receiving a `prepare_ok` for any in-flight prepare, pause `command=commit` heartbeats.
This, in turn, will allow the backups to kick off an SVC and potentially a view-change, depending on the remainder of the network's health.

`ping`/`pong` no longer requires an attached `view`.

`commit` now includes a monotonic timestamp, which is used to discard old/duplicated heartbeats.

It is essential that the primary step down if it cannot make progress.
It is possible for all packets _except_ `command=prepare` to succeed (`command=prepare` tend to be larger, so are more susceptible to packet loss/corruption).
In that case, heartbeats/pings would appear fine, but the cluster would be deadlocked.

sentientwaffle committed:
VSR: Retry non-head prepare of pipeline

Suggested via https://github.com/tigerbeetledb/tigerbeetle/pull/458#issuecomment-1432953902

Previously `prepare_timeout` would be a noop when the pipeline's head already has a quorum, but is committing.
Now it will retry the prepare of the first pipeline entry without its quorum.

This has a bonus benefit that didn't occur to me until after I implemented it:
It isn't in `main` yet, but primaries will soon "abdicate" if they haven't received a `prepare_ok` recently.
So ensuring that we don't "pause" prepares just because we are busy committing reduces the chances that the primary will abdicate when it shouldn't

There is a `replica_count = 1` workaround since `replica_count = 1` is not allowed to append multiple entries to the WAL concurrently.
I decided to put the condition in `on_prepare_timeout` instead of just not enabling the timeout for `replica_count=1` because the latter approach requires elaborating `assert(prepare_timeout.ticking)` in a bunch of places.

sentientwaffle committed:
VSR: prepare_timeout when ∃!ok_quorum_received

Only run `prepare_timeout` when at least one prepare in the pipeline has not finished its quorum.

Rationale:
- This makes its status exactly match the `primary_abdicate_timeout`.
- And as a bonus, the timer doesn't run when it is not needed.

sentientwaffle committed:
VSR: prepare_pipeline_pending refactor

sentientwaffle committed:
VSR: status=recovering is only replicas=1

...so there's no view-change messages.

sentientwaffle committed:
VSR: Primary sends DVC

Since DVCs are now broadcast to all replicas, the primary should send them just the same as the backups.

sentientwaffle commented on src/message_bus.zig:
```diff
@@ -806,6 +806,7 @@ fn MessageBusType(comptime process_type: vsr.ProcessType) type {
                         log.info("connection from replica {}", .{connection.peer.replica});
                     },
                     .client => {
+                        assert(header.client != 0);
```
This was already true before this PR. If we have determined that this connection is from a client, we expect that the header has an included client id.

sentientwaffle commented on src/simulator.zig:
```diff
@@ -143,9 +143,10 @@ pub fn main() !void {
         .replica_crash_probability = 0.00002,
         .replica_crash_stability = random.uintLessThan(u32, 1_000),
-        .replica_restart_probability = 0.0001,
+        .replica_restart_probability = 0.0002,
```
Tuning this a little since I ran into some simulator runs that took a very long time simply because they were blocked waiting for a particular replica to restart. Once state transfer is far enough along that the "whole quorum must checkpoint together" hack is removed, we'll be able to test much longer crashes.

sentientwaffle commented on src/simulator.zig:
```diff
@@ -268,9 +269,9 @@ pub const Simulator = struct {
     requests_idle: bool = false,
 
     pub fn init(allocator: std.mem.Allocator, random: std.rand.Random, options: Options) !Simulator {
-        assert(options.replica_crash_probability < 1.0);
+        assert(options.replica_crash_probability < 100.0);
         assert(options.replica_crash_probability >= 0.0);
-        assert(options.replica_restart_probability < 1.0);
+        assert(options.replica_restart_probability < 100.0);
```
Bug, though all of the percentages were below 1 anyway.

matklad replied:
Hm, that feels like a confusing name. If I see `.replica_restart_probability = 0.0002`, I'd totally interpret that as probability (number between 0 and 1) rather than percentage. That is, for me at least, probability is very strongly a number between 0 and 1.

What's our overall plan here? I think I've heard that we want to get rid of floats altogether...

For me, one of the two would makes sense: 

* specify float probabilities everywhere (so, `crash_probability = 0.000002`)
* specify integers with fixed point semantics (so, `crash_percent = 1`, except that we need something with much higher denominator. Maybe `crash_per_million = 2`)

Somewhat out of scope for the PR i guess

sentientwaffle replied:
#2 is the plan (integers everywhere), but admittedly not a particularly high priority.
Agreed that `_probability < 100` is surprising (hence this bug), I think it ended up this way because the integer probabilities use 100 as a denominator, so the floats ended up the same.

sentientwaffle replied:
For context, we want to get rid of floats b/c floats aren't necessarily deterministic between different CPUs.

sentientwaffle commented on src/vsr.zig:
```diff
@@ -491,7 +526,7 @@ pub const Header = extern struct {
         if (self.request != 0) return "request != 0";
         if (self.op != 0) return "op != 0";
-        if (self.timestamp != 0) return "timestamp != 0";
+        if (self.timestamp == 0) return "timestamp == 0";
```
Since `commit` messages indicate to backups that the primary is alive, they use `timestamp` to discard old (or repeated) heartbeats.

sentientwaffle commented on src/vsr/journal.zig:
```diff
@@ -1643,6 +1643,8 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
             assert(journal.has_dirty(message.header));
 
             const write = journal.writes.acquire() orelse {
+                assert(replica.replica_count > 1);
```
This was already true, just adding an assert because I got tripped up on this at one point.

`replica_count=1` serializes its WAL writes writes (unlike all other cluster sizes which allow concurrent writes).
The reason is that otherwise `replica_count=1` risk forking the log if they ever lose a write (e.g. in the middle of a pipeline) followed by a crash/restart. Of course cluster-of-1 can't recover from a lost write, but this downgrades it from a correctness issue to an availability issue.

An alternate approach would be: Don't have special cases to serialize WAL writes in the replica, and instead hard-code `replica_count=1`'s pipeline to be 1 entry. But that would make a comptime parameter (pipeline size) depend on a runtime parameter (cluster size), which is annoying, and also I hadn't thought of that approach until just now...

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -5336,24 +5510,27 @@ pub fn ReplicaType(
                 self.repair_timeout.start();
             }
 
+            self.heartbeat_timestamp = 0;
```
Reset any time the primary changes since different replicas have different clocks.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4943,11 +5104,9 @@ pub fn ReplicaType(
-                },
+                // The replica's view hasn't been set yet.
+                // It will be set shortly, when we transition to normal status.
+                .recovering => assert(self.view == 0),
```
Random though: many of these "replica_count=1 never changes views" assertions won't be valid once we have reconfiguration, since you could reconfigure down to replicas=1 then back up. (Not that someone should ever do that. :fearful:)

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4929,11 +5094,7 @@ pub fn ReplicaType(
-                {
-                    self.send_do_view_change();
-                }
+                if (!self.do_view_change_quorum) self.send_do_view_change();
```
DVCs used to be from replica → primary, but now they are broadcast to all other replicas.
The primary uses them for the same purpose as before. Backups use them to trigger their own DVCs.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4546,7 +4706,14 @@ pub fn ReplicaType(
+                self.do_view_change_from_all_replicas[self.replica] == null)
+            {
+                self.send_message_to_replica(self.replica, message);
+                defer self.flush_loopback_queue();
```
The `defer` here doesn't do anything, it is more of a hint that `flush_loopback_queue` shouldn't have anything after it. (We do the same elsewhere for `flush_loopback_queue`).

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4506,28 +4666,28 @@ pub fn ReplicaType(
+
+            self.send_header_to_other_replicas(header);
+
+            if (!self.start_view_change_from_all_replicas.isSet(self.replica)) {
+                self.send_header_to_replica(self.replica, header);
+                defer self.flush_loopback_queue();
+            }
```
Why send ourselves an SVC message rather than simply flipping the bit in the quorum count?
This may be the last SVC needed to complete our quorum — using the regular message path means there's only one code path for quorum completion.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4412,13 +4572,13 @@ pub fn ReplicaType(
         }
 
         fn reset_quorum_nack_prepare(self: *Self) void {
+            assert(!self.nack_prepare_from_other_replicas.isSet(self.replica));
```
This was already true; we never send a nack message to ourselves.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -2288,10 +2411,6 @@ pub fn ReplicaType(
-                } else {
-                    // When the pipeline is empty, stop the prepare timeout.
-                    // The timeout will be restarted when another entry arrives for the pipeline.
-                    self.prepare_timeout.stop();
```
This timeout used to run when the pipeline was non-empty.
Now this timeout runs when the pipeline is non-empty _and_ at least one prepare in the pipeline is awaiting a quorum of acks (i.e. `!prepare.ok_quorum_received`).

sentientwaffle committed:
VSR: Tighten create_message_from_header assertion

Replica never sends `.ping_client`.
Replica does send `.pong_client`, but it includes `header.view == replica.view`, so the assert passes.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -578,16 +609,31 @@ pub fn ReplicaType(
+                    .name = "primary_abdicate_timeout",
                     .id = replica_index,
-                    .after = 100,
+                    .after = 1000,
```
This is a long timeout, but since the cost of a false positive is high I think that's ok.

matklad commented on src/testing/packet_simulator.zig:
```diff
@@ -64,8 +64,11 @@ pub const PartitionMode = enum {
+
+    /// Isolates exactly one replica (asymmetric: send-only).
+    isolate_single_sender,
 };
```
assymetric partitons make sense for every mode except `none`, so perhaps we a separate `assymetric_partitions: bool` param?

matklad commented on src/vsr.zig:
```diff
@@ -348,15 +354,13 @@ pub const Header = extern struct {
-            if (self.replica != 0) return "replica != 0";
-            if (self.op != 0) return "op != 0";
-        }
+        if (self.size != @sizeOf(Header)) return "size != @sizeOf(Header)";
```
:+1:

matklad replied:
Thinking about this ... if we assert that we have no body, shouldn't we also assert about `checksum_body`?

matklad commented on src/vsr.zig:
```diff
@@ -367,10 +371,40 @@ pub const Header = extern struct {
+        if (self.client != 0) return "client != 0";
+        if (self.context != 0) return "context != 0";
+        if (self.request != 0) return "request != 0";
+        if (self.op != 0) return "op != 0";
```
A bit annoying that we can't `assert (self.view > 0)` here. Maybe we should start with view 1 ?

matklad commented on src/vsr.zig:
```diff
@@ -687,7 +714,7 @@ pub const Timeout = struct {
     /// It's important to check that when fired() is acted on that the timeout is stopped/started,
     /// otherwise further ticks around the event loop may trigger a thundering herd of messages.
-    pub fn fired(self: *Timeout) bool {
+    pub fn fired(self: *const Timeout) bool {
```
Would it be correct to check `.fired()` two times in a row? The code looks like it wants to set some flag or something for fired timers...

sentientwaffle replied:
:thinking: If it was `.fire()` I would agree, but since it is `.fired()` it seems ok to call it multiple times in the same tick.

matklad replied:
Yeah, so we can do either of: 

* keep name, change to `*const`
* keep `*Timeout`, change name to `fire`, add assert

The comment above makes me think that maybe the second one is the right approach. 

:thinking: or we can even have both functions, asserting`.fire()` and  const `.is_firing()`, with the expectation that most call-sites want `.fire`, but some specific ones need `is_firing()` (and guard against thundering herd in some way. )

matklad replied:
Also perhaps out of scope for the PR, as we don't have assert today, so it's not clear at a glance if "call multiple times in a tick" is happenning or not in practiec.

sentientwaffle replied:
Since `fire()`/`fired()` (either one) _must_ be called at _every_ tick (otherwise we [panic](https://github.com/tigerbeetledb/tigerbeetle/blob/bfbcd126353d0161b8fbcd14260759c39709071a/src/vsr.zig#L693-L696)), it seems like `tick()` could return a "did it fire" boolean. i.e. instead of:

```zig
            self.ping_timeout.tick();
...
            if (self.ping_timeout.fired()) self.on_ping_timeout();
```
it would be


```zig
            if (self.ping_timeout.tick()) self.on_ping_timeout();
```

matklad replied:
Ah, we _only_ ever call `.tick` and `.fire()` in that small section of code, this is so much better than what I imagined happening. I think that just changing to `*const` is best then. I _like_ the current setup where we first tick all timers, and then check each one for `fired`, because it's more order-independent: logically, we advance all timers at the same time, there no transitionary states where some timers were advanced, and some were not.

matklad commented on src/vsr/client.zig:
```diff
@@ -274,7 +274,7 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
 
         fn on_pong(self: *Self, pong: *const Message) void {
-            assert(pong.header.command == .pong);
+            assert(pong.header.command == .pong_client);
```
`fn on_pong_client` then?

matklad replied:
relatedly, while understanding view change protocol, the `start_view` / `start_view_change` naming tripped me so many times. It's too easy to search for `start_view_change_something`  and end up looking at the `start_view` code instead. 

If we follow through with our MSB naming, it probably should be:

```
viewchange_start
viewchange_do
view_start
request_view_start.
```

matklad replied:
So, yeah, given that this is client, I think `on_pong` is ok. Though, I'd still prefer `on_pong_client` perhaps for consistency.

matklad commented on src/vsr/client.zig:
```diff
@@ -390,7 +390,7 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
 
             const ping = Header{
-                .command = .ping,
+                .command = .ping_client,
```
Wait, this doesn't make sense. The command is `.ping_replica`, but we `send_header_to_replicas`. Either this should remain `.ping`, or I completely misunderstand the intended semantics here.

sentientwaffle replied:
My intent is:

- `ping_client` is client -> replica
- `pong_client` is replica -> client
- `ping` is replica -> replica
- `pong` is replica -> replica

So `_client` denotes that the client is on one end of the message, not necessarily that they are the receiver.

matklad replied:
:facepalm: I was thinking this is `replica.zig`, not `client.zig`.

matklad commented on src/vsr/replica.zig:
```diff
@@ -208,18 +208,27 @@ pub fn ReplicaType(
+        /// The last timestamp received on a commit heartbeat.
+        /// Used to discard delayed heartbeat messages.
+        /// (status=normal backup)
+        heartbeat_timestamp: u64 = 0,
```
Could you say in the comment which clock is that? I can see at least two choices: 

* replica's local time
* our cluster time

(and I don't yet fully understand cluster time)

sentientwaffle replied:
:+1: I'll add a comment. Fyi it is the primary's monotonic clock — all we are about here is the ability to discard old/duplicate messages.

matklad replied:
Yeah, I think a `(view, primary.monotonic)` is enough for discarding. We need view though, to avoid comparing clocks of different primaries.

sentientwaffle replied:
Yeah, we ignore any `command=commit` messages from a different view.

matklad commented on src/vsr/replica.zig:
```diff
@@ -230,28 +239,50 @@ pub fn ReplicaType(
         /// The number of ticks before a primary or backup broadcasts a ping to other replicas.
         /// TODO Explain why we need this (MessageBus handshaking, leapfrogging faulty replicas,
         /// deciding whether starting a view change would be detrimental under some network partitions).
+        /// (status=normal replicas)
```
Love this additions!

matklad commented on src/vsr/replica.zig:
```diff
@@ -230,28 +239,50 @@ pub fn ReplicaType(
-        /// The number of ticks before resending a `start_view_change` or `do_view_change` message:
+        /// The number of ticks before resending a `do_view_change` message:
+        /// (status=view-change)
         view_change_message_timeout: Timeout,
```
Can perhaps rename to `do_view_change_message_timeout` now?

matklad commented on src/vsr/replica.zig:
```diff
@@ -783,65 +837,69 @@ pub fn ReplicaType(
+            assert(message.header.command == .ping);
             if (self.status != .normal and self.status != .view_change) return;
 
-            assert(self.status == .normal or self.status == .view_change);
```
One benefit of this assert is that it states the positive version of the invariant. That's why, when I write early return if's, I usually use `if (!(invariant))` form, even if pulling `!` inside can make the expression simpler. It's easy to read it as "`invariant` holds" then.

matklad commented on src/vsr/replica.zig:
```diff
@@ -783,65 +837,69 @@ pub fn ReplicaType(
         fn on_pong(self: *Self, message: *const Message) void {
-            if (message.header.client > 0) return;
+            assert(message.header.command == .pong);
             if (message.header.replica == self.replica) return;
```
Hm, reading `on_pong` makes me appreciate ` // TODO Drop pings that were not addressed to us.` comment more. It seems that a misdirected ping/pong could make our clock go askew. I guess we want to stuff target replica into `.context`?

sentientwaffle replied:
> It seems that a misdirected ping/pong could make our clock go askew.

How so? The `pong` includes the original timestamp we sent (in the ping), which we use when synchronizing the clock.

matklad replied:
if the `pong` is misdirected, than original timestamp is not what _we_ sent, but rather what some other replica sent.

sentientwaffle replied:
Ooh, I see

sentientwaffle committed:
PacketSimulator: PartitionSymmetry

matklad commented on src/vsr/replica.zig:
```diff
@@ -1746,13 +1820,39 @@ pub fn ReplicaType(
+            assert(self.status == .normal);
+            assert(self.primary());
+            assert(self.primary_pipeline_pending() != null);
+            self.primary_abdicate_timeout.reset();
```
Feels a bit weird that we do a `.reset()` here rather than a `.stop()`, as any further ticking is useless. At the same time, this makes it much easier to formulate when the timeout should be ticking. I think I like the current approach more, even if it is wasteful in some sense.

sentientwaffle committed:
VSR: Count views from 1

This allows asserting `view ≠ 0` in messages that expect a view number.
Previously they couldn't check for the absence of a view number since `0` was a valid view number.

matklad commented on src/vsr/replica.zig:
```diff
@@ -5336,24 +5509,27 @@ pub fn ReplicaType(
             }
 
+            self.heartbeat_timestamp = 0;
+            self.primary_abdicating = false;
```
This can be assert I think? `primary_abdicating` can't be true if `.status = .view_change`, and it was view_chaeg.

sentientwaffle committed:
VSR: Assert checksum_body of no-body messages

sentientwaffle committed:
VSR: Document heartbeat_timestamp's clock type

sentientwaffle committed:
VSR: Rename {,do_}view_change_message_timeout

To better match `commit_message_timeout`/`start_view_change_message_timeout`.

sentientwaffle committed:
VSR: Add back invariant check

sentientwaffle committed:
VSR/Client: Rename on_ping{,_client}

matklad commented on src/vsr/replica.zig:
```diff
@@ -1768,34 +1868,65 @@ pub fn ReplicaType(
                 .replica = self.replica,
                 .view = self.view,
                 .commit = self.commit_max,
+                .timestamp = self.clock.monotonic(),
```
update comment on `fn monotonic`.

matklad commented on src/vsr/replica.zig:
```diff
@@ -1070,6 +1141,13 @@ pub fn ReplicaType(
             assert(message.header.view == self.view);
             assert(message.header.replica == self.primary_index(message.header.view));
 
+            // Old/duplicate heartbeats don't count.
```
add `assert(message.header.command == .commit)` at the start of the function

matklad replied:
Actaually, I think we don't do this for other functions at the moment, so might start doing that (eg, in `on_start_view_change`)

sentientwaffle committed:
VSR: Improve assertions

sentientwaffle committed:
VSR: Fix client builds

matklad commented on src/vsr/replica.zig:
```diff
@@ -1154,45 +1238,41 @@ pub fn ReplicaType(
-            // number.
-            self.send_do_view_change();
-            defer self.flush_loopback_queue();
+            self.transition_to_view_change_status(self.view + 1);
```
Maybe add 

```
assert(self.start_view_change_from_all_replicas.count() == 0);
```

at the end? As a local proof that `assert(count <= threshold);` can't fire

matklad commented on src/vsr/replica.zig:
```diff
@@ -1070,6 +1148,13 @@ pub fn ReplicaType(
             assert(message.header.view == self.view);
```
```
            if (message.header.view > self.view) {
                log.debug("{}: on_commit: ignoring (newer view)", .{self.replica});
                return;
            }
```

Is this actually possible? We do a view_jump on commit, so our view should not be less?

sentientwaffle replied:
When handling `commit`/`prepare` messages, `view_jump()` does not update the `view` immediately, it just sends a `request_start_view`.

matklad replied:
Ah, yes, it is possible: view_jump won't actually jump on commit, it'll just request `start_view`.

sentientwaffle committed:
VSR: Assert message header commands

matklad commented on src/vsr/replica.zig:
```diff
@@ -230,28 +242,50 @@ pub fn ReplicaType(
+
+        /// The number of ticks before resetting the SVC quorum.
+        /// (status=normal|view-change)
+        start_view_change_window_timeout: Timeout,
```
Hm, why do we need this timeout, and what's the overall flow here? If I understand the implementation correctly, this timeout always ticks, and just resets the quorum. This ... feels weird. 

In particular, the timeout here seems to be orthogonal to the process we use so start sending SVCs. This means we may accidentally trip this timeout just before we collected an SVC quorum? 

I would maybe expect us to just reset the quorum on commit?

sentientwaffle replied:
Suppose:
1. There is a temporary network issue
2. We collect a few SVCs
3. The network issue heals
4. (much time passes)
5. There is another network issue.
6. At this point we want to start collecting SVCs, but we want to start from scratch, not from wherever our old quorum left off.

matklad replied:
Hm, and `The network issue heals` means "we received a commit" I think? So that's the place to reset the quorum? I agree that we should reset the quorum at some point, I don't understand why we use timer for that.

sentientwaffle replied:
Except it may be the the _other_ replicas received a commit (so they stopped sending SVCs, they're healthy), but we ourselves didn't see a commit — all we know is we stopped seeing SVCs.

i.e. we are partitioned from the primary (not receiving `commit`s). Some other replicas were temporarily partitioned from the primary, but recovered. Their SVCs are stale and shouldn't count towards a quorum unless they are recent enough to reflect those other replica's actual (current) view of the network.

matklad replied:
Oh, I think I see what you mean -- we don't want SVCs to be too stale. 

Yeah, it semes that we need some timeout here, but perhaps we should start it after sending/receiving the first SVC, rather than when transitioning to normal?

The case I worry about is that everyone notices that the primary is dead, everyone sends an SVC, we receive SVCs and it just so happens that this timeout triggers, and we clear SVCs, though we received all of them like 1 ms ago.

sentientwaffle replied:
What about `self.start_view_change_window_timeout.reset();` any time we receive any SVC?

matklad replied:
I think that’s ok, although doesn’t yet feel like the perfect soln (to be fair, my proposal to start the timer on SVC reciting also feels adhoc)

sentientwaffle replied:
I think the "start timer on SVC" solution reflects the concept better, but I like keeping the invariants of "when does the timer run" from getting too complicated, even when that means putting an early-return into to timeout handler function.

sentientwaffle replied:
Maybe ideally each SVC would expire independently, but I don't want to manage that many timers.

matklad replied:
Thinking more, I don’t like “reset on any SVC”: if two replicas are partitioned together, they reset each other’s timers, and keep stale SVCs from other replicas.

> Maybe ideally each SVC would expire independently, but I don't want to manage that many timers.

Yeah, that would be inelegant I would think!

> better, but I like keeping the invariants of "when does the timer run" from getting too complicated,

Uhu! At the same time, the current invariant is “timer always runs”, which has the drawback that it doesn’t give you any extra space for sanity checking or what not.

I think what I am suggesting is “timer is ticking if SVC quorum is non-empty”, which feels rather nice.

sentientwaffle replied:
Implemented, though instead of “timer is ticking if SVC quorum is non-empty” I went with "timer is ticking if SVC quorum has an SVC from someone _other than ourself_". The staleless/timeout isn't applicable to our own SVC, since we can keep that up-to-date more accurately based on the heartbeats.

sentientwaffle committed:
VSR: root prepare primary_index(view)≠replica

Counting the view from `1` allows additional validation of VSR headers.

But there is a gotcha:

The root prepare also needs a view of 1. But then what is its replica?
We want it to be 0 because that is the only replica that exists in all cluster sizes.
But then primary_index(root.view) ≠ root.replica, which holds in all other cases.

I think verifying header `view`s is still worth it, though.

I don't want the root prepare to depend on the `replica_count` since `replica_count` can vary (reconfiguration)

sentientwaffle committed:
VSR: Run SVC-window timeout only when we have SVCs

Specifically when we have any SVC from *another* replica, since the timeout doesn't impact the SVC from ourself.

matklad commented on src/vsr/replica.zig:
```diff
@@ -1701,10 +1794,15 @@ pub fn ReplicaType(
             }
 
             if (waiting_len == 0) {
```
Hm, given that we explicitly select prepare as `const prepare = self.primary_pipeline_pending().?;`, wouldn't this branch be dead code now?

sentientwaffle replied:
Aha I thought that too originally, but `waiting_len` only counts the other replicas waiting.
When `self.quorum_replication == self.replica_count`, we might be waiting for only ourself to prepare (i.e. write to the WAL).
(The assert is currently commented out because of a kludge that can be removed when state transfer is ready).

matklad commented on src/vsr/replica.zig:
```diff
@@ -1154,45 +1240,47 @@ pub fn ReplicaType(
 
-            assert(count == threshold);
-            assert(!self.start_view_change_from_other_replicas.isSet(self.replica));
+            if (self.replica != message.header.replica) {
```
```suggestion
            if (self.replica != message.header.replica and !self.start_view_change_window_timeout.ticking) {
```

I think we don't want to reset an already ticking window?

matklad commented on src/testing/packet_simulator.zig:
```diff
@@ -64,10 +66,12 @@ pub const PartitionMode = enum {
     uniform_partition,
 
-    /// Isolates exactly one replica.
+    /// Isolates exactly one replica (symmetric).
```
```suggestion
    /// Isolates exactly one replica.
```

matklad commented on src/testing/packet_simulator.zig:
```diff
@@ -257,8 +262,12 @@ pub fn PacketSimulatorType(comptime Packet: type) type {
+                        to >= self.options.replica_count or
+                        partition[from] == partition[to] or
+                        (self.options.partition_symmetry == .asymmetric and
+                        partition[from] == random.boolean());
```
This simplifies to 

```suggestion
                        (self.options.partition_symmetry == .asymmetric and random.boolean());
```

?

Perhaps we want to do rather

```
const assymetric_partition_side = random.boolean()
while (to < self.node_count()) : (to += 1) { 
   ...
  (self.options.partition_symmetry == .asymmetric and
                        partition[from] == assymetric_partition_side);
}
```

matklad commented on src/vsr.zig:
```diff
@@ -347,37 +355,71 @@ pub const Header = extern struct {
 
     fn invalid_ping(self: *const Header) ?[]const u8 {
         assert(self.command == .ping);
+        if (self.checksum_body != checksum_body_empty) return "checksum_body != expected";
```
Let's place that right next to `if (self.size != @sizeOf(Header))`: I think we always want both asserts together?

sentientwaffle replied:
:+1: Yeah, they are always accompanying each other.

(Fyi my motivation for putting them at the front was that most of the field validations are (roughly) ordered to mirror the `Header` struct layout).

matklad commented on src/vsr/client.zig:
```diff
@@ -66,7 +66,7 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
         /// The highest view number seen by the client in messages exchanged with the cluster.
         /// Used to locate the current primary, and provide more information to a partitioned primary.
-        view: u32 = 0,
+        view: u32 = 1,
```
Let's motivate why we start with 1 rather than 0, something like: "we count views from 1 to distinguish cases where the view is not set"

matklad replied:
Ah, it's client.zig again :facepalm: I think we should motivate in `root_prepare`, and here just refer to that.

sentientwaffle replied:
I'm switching back to `view=0` at joran's request, so it's moot.

matklad commented on src/vsr/replica.zig:
```diff
@@ -783,65 +840,71 @@ pub fn ReplicaType(
+            });
+        }
+
+        /// pong_client messages are only sent to clients, but a network fault can redirect one.
```
We have `// A replica should never handle misdirected messages intended for a client:` branch in the main dispatch loop for this case already

matklad commented on src/vsr/replica.zig:
```diff
@@ -1020,6 +1083,13 @@ pub fn ReplicaType(
+            if (!prepare.ok_from_all_replicas.isSet(message.header.replica)) {
+                self.primary_abdicating = false;
+                if (!prepare.ok_quorum_received) {
+                    self.primary_abdicate_timeout.reset();
```
Curiously, this seems to be the single place where `prepare_timeout` and `primary_abdicate_timeout` behave differently. This makes me think this two things: 

* perhaps we can cast `abdiacte_timeout` as just `prepare_timeout` firing several times? Eg, prepare_timeouts.attempts % 4 == 0 or something
* I might be missing something, but it seems we never reset prepare_timeout? This feels like a bug. Suppose cluster is in a steady state with a stable primary, whose pipeline is always non-empty. As we never reset the timeout, we will accumulate attempts. As we do exponential backoffs, our actual prepare tiemout would grow, even everything is ok. So, when we actually do drop a prepare, we'll take a long time to notice that.

sentientwaffle replied:
> perhaps we can cast abdiacte_timeout as just prepare_timeout firing several times? Eg, prepare_timeouts.attempts % 4 == 0 or something

Hmm I'm on the fence here.

Currently we reset whenever any "new" `prepare_ok` arrives, which is a bit more forgiving than using `prepare_timeout.attempts % n == 0`.

e.g. with `attempts % n` and a large pipeline, we might receive a bunch of later prepares while we are blocked (and retrying) an earlier one. If the replicas are successfully acking any prepares, shouldn't that be good enough?

> I might be missing something, but it seems we never reset prepare_timeout?
:+1: I'll add a reset in `on_prepare_ok` when we acquire a quorum (i.e. advance to retrying the next prepare in the pipeline).

matklad replied:
>Hmm I'm on the fence here.

One tie breaker is that, with a separate timeout, it would be easier to _change_ the logic if for some reason it has to become different.

sentientwaffle replied:
Alright, I'll leave it as-is for now.

sentientwaffle committed:
Revert "VSR: Count views from 1"

This reverts commit 98f10e085a4d5b5815d4bbd19d0f1381f7da9088.

sentientwaffle committed:
VSR: Add back reverted assert

sentientwaffle committed:
VSR: Reorganize checksum_body validation checks

sentientwaffle committed:
PacketSimulator: Fix asymmetric partitions

sentientwaffle committed:
VSR: Remove pong_client handling

sentientwaffle committed:
VSR: Fix timeouts

matklad commented on src/vsr.zig:
```diff
@@ -651,6 +701,9 @@ pub const Header = extern struct {
             .cluster = cluster,
             .size = @sizeOf(Header),
+            // Note the root prepare is the only case where:
+            //   primary_index(prepare.view) ≠ prepare.replica
```
obsolete comment?

sentientwaffle replied:
Ah, forgot to revert that second commit.

sentientwaffle committed:
Revert "VSR: root prepare primary_index(view)≠replica"

This reverts commit 8eb1a69e2cfc72bfd72cdb5f44820b0ee771c7c0.

matklad commented on src/vsr/replica.zig:
```diff
@@ -1035,6 +1093,16 @@ pub fn ReplicaType(
+                self.prepare_timeout.stop();
+                self.primary_abdicate_timeout.stop();
+            } else {
+                self.prepare_timeout.reset();
```
Maybe only reset if this is `prepare_ok` for the pending prepare (rather than for any following prepare in the pipeline)?

sentientwaffle replied:
Ooh, good catch.

sentientwaffle committed:
VSR: Reset pipeline when advancing head-prepare

matklad commented on src/vsr/replica.zig:
```diff
@@ -1205,21 +1285,21 @@ pub fn ReplicaType(
         /// ⟨start_view v, l, n, k⟩ messages to the other replicas, where l is the new log, n is the
         /// op number, and k is the commit number.
         fn on_do_view_change(self: *Self, message: *Message) void {
```
Update the comment that DVC serves two purposes: 

* for primary, to become part of DVC quorum
* for backup, to force view_jump.

sentientwaffle committed:
VSR: Mention why backups receive DVCs

matklad commented on src/vsr/replica.zig:
```diff
@@ -1035,6 +1093,18 @@ pub fn ReplicaType(
+                self.prepare_timeout.stop();
+                self.primary_abdicate_timeout.stop();
+            } else {
+                if (prepare == self.pipeline.queue.prepare_queue.head_ptr().?) {
```
Hm, I was thinking more about `self.primary_pipeline_pending().?` rather than literally the first one (the headone might be waitin for disk, and the secodn might be waiting for quorum)

matklad replied:
Though, I think this should also be fine.

matklad commented:
LGTM, though maybe we can be a bit more precise here: https://github.com/tigerbeetledb/tigerbeetle/pull/476#discussion_r1114572962


bors d+

sentientwaffle committed:
VSR: Reset prepare timeout when advancing pending-pipeline-prepare

sentientwaffle commented:
bors r+
